src/renderer/atlas/AtlasEngine.r.cpp

```diff
@@ -294,31 +267,29 @@ void AtlasEngine::_drawGlyph(const AtlasQueueItem& item) const
         _r.d2dRenderTarget->SetTextAntialiasMode(coloredGlyph ? D2D1_TEXT_ANTIALIAS_MODE_GRAYSCALE : D2D1_TEXT_ANTIALIAS_MODE_CLEARTYPE);
     }
 
-    _r.d2dRenderTarget->BeginDraw();
-    // We could call
-    //   _r.d2dRenderTarget->PushAxisAlignedClip(&rect, D2D1_ANTIALIAS_MODE_ALIASED);
-    // now to reduce the surface that needs to be cleared, but this decreases
-    // performance by 10% (tested using debugGlyphGenerationPerformance).
-    _r.d2dRenderTarget->Clear();
-    _r.d2dRenderTarget->DrawTextLayout({}, textLayout.get(), _r.brush.get(), options);
-    THROW_IF_FAILED(_r.d2dRenderTarget->EndDraw());
-
     for (u32 i = 0; i < cells; ++i)
     {
-        // Specifying NO_OVERWRITE means that the system can assume that existing references to the surface that
-        // may be in flight on the GPU will not be affected by the update, so the copy can proceed immediately
-        // (avoiding either a batch flush or the system maintaining multiple copies of the resource behind the scenes).
-        //
-        // Since our shader only draws whatever is in the atlas, and since we don't replace glyph tiles that are in use,
-        // we can safely (?) tell the GPU that we don't overwrite parts of our atlas that are in use.
-        _copyScratchpadTile(i, coords[i], D3D11_COPY_NO_OVERWRITE);
+        const auto coord = coords[i];
+
+        D2D1_RECT_F rect;
+        rect.left = static_cast<float>(coord.x) * static_cast<float>(USER_DEFAULT_SCREEN_DPI) / static_cast<float>(_r.dpi);
+        rect.top = static_cast<float>(coord.y) * static_cast<float>(USER_DEFAULT_SCREEN_DPI) / static_cast<float>(_r.dpi);

```

Comment:
This bit is a bit unrefined, but in the grand scheme of things isn't the biggest issue about this class. 😅
I'd much prefer if D2D would be DPI-unaware so that I can set `rect` and `origin` coordinates straight to pixel coordinates, but IIRC that's problematic for AA when using the non-text primitives (for instance when drawing the cursor). Might be worthwhile to investigate at another time.